src/util/pg.py

```diff
@@ -97,27 +97,55 @@ def _parallel_execute_threaded(cr, queries, logger=_logger):
         max_workers = min(get_max_workers(), len(queries))
         cursor = db_connect(cr.dbname).cursor
 
+        PG_CONCURRENCY_ERRORS_TO_RETRY = {
+            errorcodes.LOCK_NOT_AVAILABLE,
+            errorcodes.SERIALIZATION_FAILURE,
+            errorcodes.DEADLOCK_DETECTED,
+        }
+
         def execute(query):
+            """
+            If a concurrency-related issue is raised, the cursor is rolled back and the query is returned.
+            Otherwise the number of rows affected is returned.
+            """
             with cursor() as cr:
-                cr.execute(query)
-                cnt = cr.rowcount
-                cr.commit()
-                return cnt
+                try:
+                    cr.execute(query)
+                    cnt = cr.rowcount
+                    cr.commit()
+                except psycopg2.OperationalError as exc:
+                    if exc.pgcode not in PG_CONCURRENCY_ERRORS_TO_RETRY:
+                        raise
+                    cr.rollback()
+                    return query
```

Comment:
I don't like this mixed returns. I'd just define a `conflict_queries = []` outside `execute` and add queries there when they fail. The use that  `_parallel_execute_serial(cr, conflict_queries, ...)`

Reply:
> That idea doesn't sound thread safe.

I think it is python-thread safe (due to the GIL). Still, it's probably safer not to do that.

Reply:
You can use a synchronized queue, but well maybe not worth the hassle.

Reply:
Yeah, I'm not sure it is. I will do a fixup commit, then we choose.

Reply:
> I think it is python-thread safe (due to the GIL)

That's an implementation detail I would not make assumptions or rely on. You could probably get away with a bit of manual locking.

```python
from threading import Lock

def execute():
    ...
    with lock:
        conflict_queries.append(query)

conflict_queries = []
lock = Lock()

with ThreadPoolExecutor(max_workers=max_workers) as executor:
....
```

Reply:
Update: I think that the synchronized queue makes for a nice solution here.